# **ЗАДАНИЕ.**
<br>
Реализуйте первичный анализ данных о фильмах. Для этого используйте набор данных <b>Dataset1.csv</b>. При выполнении задания выполните следующее.
<br>
1. Cформулируйте три гипотезы и проверьте правильность своих предположений, подробно поясняя как в приведенном выше проекте.
<br>
Чтобы иметь возможность сравнивать фильмы разных лет, необходимо будет преобразовать валовую и бюджетную стоимость в реальные суммы в долларах с точки зрения покупательной способности 2016 года. Для этого используйте индекс потребительских цен (Consumer Price Index, CPI) с поправкой на инфляцию (https://www.usinflationcalculator.com/inflation/consumer-price-index-and-annual-percent-changes-from-1913-to-2008/).
<br>
Посчитайте значения CPI за каждый год с 1913 по 2017 гг включительно. Используя <i>Калькулятор инфляции в США</i> в качестве источника и постройте DataFrame.
<br>
Пример кода представлен в файле <b>Задание.pdf</b>
Полученную таблицу, содержащую CPI, экспортируйте в файл формата *.csv с именем  <font color=blue>Lab1_Фамилия_CPI</font>, например, <font color=blue><i>Lab1_Ivanov_CPI.csv</i></font>
<br>
<font color=green><u>Самостоятельно изучите команду экспорта датафрейма в файл CSV</u></font>:
https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_csv.html
<br><br>
2. Обработанный датафрейм экспортируйте в файл формата *.csv с именем  <font color=blue>Lab1_Фамилия_группа</font>, например, <font color=blue><i>Lab1_Ivanov_MO-221.csv</i></font>


In [1]:
import pandas as pd

Получим информацию о датасете

In [2]:
df =  pd.read_csv(f'C:/Users/User/ml-labs/data/Dataset1.csv')
df.head (15)

color      Director_Name  num_Critic_for_reviews  duration  \
0    Color      James Cameron                   723.0     178.0   
1   Colour     Gore Verbinski                   302.0     169.0   
2   Colour         Sam Mendes                   602.0     148.0   
3    Color  Christopher Nolan                   813.0     164.0   
4      NaN        Doug Walker                     NaN       NaN   
5   Colour     Andrew Stanton                   462.0     132.0   
6    Color          Sam Raimi                   392.0     156.0   
7    Color       Nathan Greno                   324.0     100.0   
8   Colour        Joss Whedon                   635.0     141.0   
9   Colour        David Yates                   375.0     153.0   
10   Color        Zack Snyder                   673.0     183.0   
11   Color       Bryan Singer                   434.0     169.0   
12   Color       Marc Forster                   403.0     106.0   
13   Color     Gore Verbinski                   313.0     151.0   
14   Color     Gore Verbinski                   450.0     150.0   

    director_Facebook_likes  actor_3_Facebook_likes       actor_2_name  \
0                       0.0                   855.0   Joel David Moore   
1                     563.0                  1000.0      Orlando Bloom   
2                       0.0                   161.0       Rory Kinnear   
3                   22000.0                 23000.0     Christian Bale   
4                     131.0                     NaN         Rob Walker   
5                     475.0                   530.0    Samantha Morton   
6                       0.0                  4000.0       James Franco   
7                      15.0                   284.0       Donna Murphy   
8                       0.0                 19000.0  Robert Downey Jr.   
9                     282.0                 10000.0   Daniel Radcliffe   
10                      0.0                  2000.0       Lauren Cohan   
11                      0.0                   903.0      Marlon Brando   
12                    395.0                   393.0    Mathieu Amalric   
13                    563.0                  1000.0      Orlando Bloom   
14                    563.0                  1000.0        Ruth Wilson   

    Actor_1_Facebook_likes        gross  \
0                   1000.0  760505847.0   
1                  40000.0  309404152.0   
2                  11000.0  200074175.0   
3                  27000.0  448130642.0   
4                    131.0          NaN   
5                    640.0   73058679.0   
6                  24000.0  336530303.0   
7                    799.0  200807262.0   
8                  26000.0  458991599.0   
9                  25000.0  301956980.0   
10                 15000.0  330249062.0   
11                 18000.0  200069408.0   
12                   451.0  168368427.0   
13                 40000.0  423032628.0   
14                 40000.0   89289910.0   

                                               genres  ...  \
0                     Action|Adventure|Fantasy|Sci-Fi  ...   
1                            Action|Adventure|Fantasy  ...   
2                           Action|Adventure|Thriller  ...   
3                                     Action|Thriller  ...   
4                                         Documentary  ...   
5                             Action|Adventure|Sci-Fi  ...   
6                            Action|Adventure|Romance  ...   
7   Adventure|Animation|Comedy|Family|Fantasy|Musi...  ...   
8                             Action|Adventure|Sci-Fi  ...   
9                    Adventure|Family|Fantasy|Mystery  ...   
10                            Action|Adventure|Sci-Fi  ...   
11                            Action|Adventure|Sci-Fi  ...   
12                                   Action|Adventure  ...   
13                           Action|Adventure|Fantasy  ...   
14                           Action|Adventure|Western  ...   

   num_user_for_reviews language  country  content_rating       b

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5043 entries, 0 to 5042
Data columns (total 28 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   color                       5025 non-null   object 
 1   Director_Name               4872 non-null   object 
 2   num_Critic_for_reviews      4927 non-null   float64
 3   duration                    4959 non-null   float64
 4   director_Facebook_likes     4872 non-null   float64
 5   actor_3_Facebook_likes      4953 non-null   float64
 6   actor_2_name                4963 non-null   object 
 7   Actor_1_Facebook_likes      4968 non-null   float64
 8   gross                       4104 non-null   float64
 9   genres                      4974 non-null   object 
 10  actor_1_name                4968 non-null   object 
 11  movie_Title                 4974 non-null   object 
 12  num_voted_users             4974 non-null   float64
 13   cast_total_facebook_likes  4974 

In [4]:
df.columns

Index(['color', 'Director_Name', 'num_Critic_for_reviews', 'duration',
       'director_Facebook_likes', 'actor_3_Facebook_likes', 'actor_2_name',
       'Actor_1_Facebook_likes', 'gross', 'genres', 'actor_1_name',
       'movie_Title', 'num_voted_users', ' cast_total_facebook_likes',
       'actor_3_name', 'facenumber_in_poster', 'plot_keywords',
       'movie_imdb_link', 'num_user_for_reviews', 'language', 'country',
       'content_rating', 'budget', 'title_year', 'actor_2_facebook_likes',
       'imdb_score', 'aspect_ratio', 'movie_facebook_likes;'],
      dtype='object')

Редактирую названия столбцов

In [5]:
df = df.rename (
    columns = {
        'movie_facebook_likes;' : 'movie_facebook_likes',
        ' cast_total_facebook_likes' : 'cast_total_facebook_likes'
    }
)

renamed_columns = {}
for column in df.columns:
    renamed_columns[column] = column.lower().strip()
    
df = df.rename(columns = renamed_columns)

In [6]:
df.columns

Index(['color', 'director_name', 'num_critic_for_reviews', 'duration',
       'director_facebook_likes', 'actor_3_facebook_likes', 'actor_2_name',
       'actor_1_facebook_likes', 'gross', 'genres', 'actor_1_name',
       'movie_title', 'num_voted_users', 'cast_total_facebook_likes',
       'actor_3_name', 'facenumber_in_poster', 'plot_keywords',
       'movie_imdb_link', 'num_user_for_reviews', 'language', 'country',
       'content_rating', 'budget', 'title_year', 'actor_2_facebook_likes',
       'imdb_score', 'aspect_ratio', 'movie_facebook_likes'],
      dtype='object')

проверю явные дубликаты и удалю их

In [7]:
df.duplicated().sum()

45

In [8]:
df = df.drop_duplicates().reset_index(drop=True)

In [9]:
df.duplicated().sum()

0

удалю неявные дубликаты

In [10]:
def replace_wrong_color (wrong_genres, correct_genre):
    for wrong_genre in wrong_genres:
        df['color'] = df['color'].replace(wrong_genre, correct_genre)

In [11]:
duplicates = ['colour','Color','Colour']
name = 'color'
replace_wrong_color(duplicates, name)

duplicates = [' Black and White',' black and white']
name = 'black and white'
replace_wrong_color(duplicates, name)

In [12]:
df.isna().sum()

color                         18
director_name                169
num_critic_for_reviews       114
duration                      83
director_facebook_likes      169
actor_3_facebook_likes        89
actor_2_name                  79
actor_1_facebook_likes        74
gross                        928
genres                        68
actor_1_name                  74
movie_title                   68
num_voted_users               68
cast_total_facebook_likes     68
actor_3_name                  89
facenumber_in_poster          81
plot_keywords                214
movie_imdb_link               68
num_user_for_reviews          86
language                      81
country                       71
content_rating               363
budget                       544
title_year                   172
actor_2_facebook_likes        79
imdb_score                    68
aspect_ratio                 386
movie_facebook_likes          68
dtype: int64

Заполню пропущенные значения, в спроках с типом string поставлю "unknown"

In [13]:
columns_to_replace_string =['color','director_name','actor_2_name','genres','actor_1_name',
    'movie_title','actor_3_name','plot_keywords','movie_imdb_link','language','country',
    'content_rating']

for row in columns_to_replace_string:
    df[row] = df[row].fillna('unknown')

В строках с типом float заполню средним по режисеру, если у режиссера одна работа то заполню нулем

In [14]:
columns_to_replace_float =['num_critic_for_reviews', 'duration','director_facebook_likes',
    'actor_3_facebook_likes','actor_1_facebook_likes','gross','num_voted_users',
    'cast_total_facebook_likes','facenumber_in_poster','num_user_for_reviews','budget',
    'title_year','actor_2_facebook_likes','imdb_score','aspect_ratio']

for column in columns_to_replace_float:
    group_by_directors=df.groupby("director_name")[column].transform("median")
    df[column]=df[column].fillna(group_by_directors)

for column in columns_to_replace_float:
    df[column] = df[column].fillna(0)


In [15]:
df.isna().sum()

color                         0
director_name                 0
num_critic_for_reviews        0
duration                      0
director_facebook_likes       0
actor_3_facebook_likes        0
actor_2_name                  0
actor_1_facebook_likes        0
gross                         0
genres                        0
actor_1_name                  0
movie_title                   0
num_voted_users               0
cast_total_facebook_likes     0
actor_3_name                  0
facenumber_in_poster          0
plot_keywords                 0
movie_imdb_link               0
num_user_for_reviews          0
language                      0
country                       0
content_rating                0
budget                        0
title_year                    0
actor_2_facebook_likes        0
imdb_score                    0
aspect_ratio                  0
movie_facebook_likes         68
dtype: int64

форматирую в  файл cvs

In [16]:
df.to_csv(r'C:/Users/User/ml-labs/data/Lab1_Karpov_FIT-221.csv', index = False)  

преобразую валовую и бюджетную стоимость в реальные сммы в долларах

In [17]:
from bs4 import BeautifulSoup
import requests

url="http://www.usinflationcalculator.com/inflation/consumer-price-index-and-annual-percent-changes-from-1913-to-2008/"

r=requests.get(url)
data=r.text
soup=BeautifulSoup(data, 'html.parser')

table=soup.find('table')
rows=table.tbody.findAll('tr')

years=[]
cpis=[]

for row in rows:
    year=row.findAll('td')[0].get_text()
    if year.isdigit() and int(year)<2017:
        years.append(int(year))
        cpis.append(float(row.findAll('td')[13].get_text()))

cpi_table=pd.DataFrame({
    "year":years,
    "avg_annual_cpi":cpis
})

cpi_table.head()

cpi_table.to_csv(r'C:/Users/User/ml-labs/data/Lab1_Karpov_CPI.csv', index = False)  

In [18]:
def get_real_value(nominal_amt,old_cpi,new_cpi):
    real_value=(nominal_amt*new_cpi)/old_cpi
    return real_value


CPI_2016=float(cpi_table[cpi_table['year']==2016]['avg_annual_cpi'])

real_domestic_gross=[]
real_budget_values=[]

for index,row in df.iterrows():
    gross=row['gross']
    budget=row['budget']
    year=row['title_year']
    cpi_series = cpi_table[cpi_table["year"] == year]["avg_annual_cpi"].values
    if len(cpi_series) > 0:
        cpi = cpi_series[0]
        real_gross = round(get_real_value(gross, cpi, CPI_2016))
        real_budget = round(get_real_value(budget, cpi, CPI_2016))
        real_domestic_gross.append(real_gross)
        real_budget_values.append(real_budget)
    else:
        real_domestic_gross.append(-1.0)
        real_budget_values.append(-1.0)
    
df['real_domestic_gross']=real_domestic_gross
df['real_budget']=real_budget_values

df.to_csv(r'C:/Users/User/ml-labs/data/Lab1_Karpov_CPI.csv')

C:\Users\User\AppData\Local\Temp\ipykernel_11468\478441109.py:6: FutureWarning: Calling float on a single element Series is deprecated and will raise a TypeError in the future. Use float(ser.iloc[0]) instead
  CPI_2016=float(cpi_table[cpi_table['year']==2016]['avg_annual_cpi'])


Проверка гипотез <br>
Гипотеза 1: В среднем в США снимают самые дорогие фильмы <br>
Гипотеза 2: Чем дороже фильм, тем выше у него оценка <br>
Гипотеза 3: Больше всего фильмов снято на английском языке <br>

проверка гипотезы 1

In [19]:
df_test_1 = pd.DataFrame(df.groupby('country')['budget'].median()).sort_values(by='budget', ascending=False)

df_test_1.head(10)



budget
country                     
Hungary         1.260000e+09
Thailand        2.000000e+08
New Line        9.000000e+07
Czech Republic  5.000000e+07
Peru            4.500000e+07
Belgium         3.850000e+07
Aruba           3.500000e+07
Libya           3.500000e+07
Chile           2.600000e+07
Japan           2.600000e+07

Оказывается у Венгрии в среднем самые высокобюджетные фильмы(???), гипотеза не подтвердилась

проверка гипотезы 2

In [20]:
df["imdb_score"].corr(df["budget"])

0.02425463986295835

Корреляция меньше 0.1, можно сказать что взаимосвязи почти нет, гипотеза не подтвердилась

проверка гипотезы 3

In [21]:
df.groupby("language")["language"].count()

language
Aboriginal       2
Arabic           5
Aramaic          1
Bosnian          1
Cantonese       11
Chinese          3
Czech            1
Danish           5
Dari             2
Dutch            4
Dzongkha         1
English       4605
Filipino         1
French          72
German          18
Greek            1
Hebrew           5
Hindi           27
Hungarian        1
Icelandic        2
Indonesian       2
Italian          9
Japanese        17
Kannada          1
Kazakh           1
Korean           7
Mandarin        22
Maya             1
Mongolian        1
Norwegian        4
Panjabi          1
Persian          4
Polish           4
Portuguese       8
Romanian         1
Russian         11
Slovenian        1
Spanish         39
Swahili          1
Swedish          5
Tamil            1
Telugu           1
Thai             3
Urdu             1
Vietnamese       1
Zulu             2
unknown         81
Name: language, dtype: int64

На английском языке снято 4605 фильмов, это абсолютное большинство, гипотеза подтвердилась